saleor/graphql/discount/dataloaders.py

```diff
@@ -330,3 +332,101 @@ def batch_load(self, keys):
         )
         promotion_map = {rule.id: promotions.get(rule.promotion_id) for rule in rules}
         return [promotion_map.get(rule_id) for rule_id in keys]
+
+
+class SaleChannelListingByPromotionIdLoader(DataLoader):
+    context_key = "sale_channel_listing_by_promotion_id"
+
+    def batch_load(self, keys):
+        from .types.sales import SaleChannelListing
+
+        def with_rules(rules):
+            rule_ids = [rule.id for item in rules for rule in item]
+
+            def with_channels(channels):
+                rule_channels = dict(zip(rule_ids, channels))
+                promotion_listing_map = defaultdict(list)
+                for promotion_id, promotion_rules in zip(keys, rules):
+                    for rule in promotion_rules:
+                        channels = rule_channels[rule.id]
+                        for channel in channels:
+                            promotion_listing_map[promotion_id].append(
+                                SaleChannelListing(
+                                    id=rule.old_channel_listing_id,
+                                    channel=channel,
+                                    discount_value=rule.reward_value,
+                                    currency=channel.currency_code,
+                                )
+                            )
+                return [promotion_listing_map[key] for key in keys]
+
+            return (
+                ChannelsByPromotionRuleIdLoader(self.context)
+                .load_many(rule_ids)
+                .then(with_channels)
+            )
+
+        return (
+            PromotionRulesByPromotionIdLoader(self.context)
+            .load_many(keys)
+            .then(with_rules)
+        )
+
+
+class PromotionRulesByPromotionIdAndChannelSlugLoader(DataLoader):
+    context_key = "promotion_rules_by_promotion_id_and_channel_slug"
+
+    def batch_load(self, keys):
+        promotion_ids = [key[0] for key in keys]
+        channel_slug = keys[0][1]
+        rules = PromotionRulesByPromotionIdLoader(self.context).load_many(promotion_ids)
+        channel = ChannelBySlugLoader(self.context).load(channel_slug)
+
+        def with_rules_and_channel(data):
+            rules, channel = data
+            rule_ids = [rule.id for item in rules for rule in item]
+            rules = PromotionRule.objects.prefetch_related("channels").filter(
+                id__in=rule_ids, channels__id=channel.id
+            )
```

Comment:
It such scenario it's no reason for using `PromotionRulesByPromotionIdLoader` and `ChannelBySlugLoader` for the start 😄 .

Probably we should use here the `through` model between `PromotionRule - Channel` to get the matching objects, instead.

Reply:
refactored